File: src/main/java/artisra/leetcode75/arraystring/CanPlaceFlowers.java
package artisra.leetcode75.arraystring;

/**
 * This class provides a solution to the "Can Place Flowers" problem on LeetCode.
 * <p>
 * Given a flowerbed represented as an array of 0s and 1s, where 0 means an empty plot and 1 means a planted flower,
 * determine if it's possible to plant a given number of new flowers without violating the rule that flowers cannot be planted
 * in adjacent plots.
 *
 * @see <a href="https://leetcode.com/problems/can-place-flowers/description/">Can Place Flowers on LeetCode</a>
 */
public class CanPlaceFlowers {

    /**
     * Determines if it's possible to plant a given number of flowers in the flowerbed.
     *
     * @param flowerBed      The array representing the flowerbed, where 0 means empty and 1 means planted.
     * @param flowersToPlant The number of flowers to plant.
     * @return `true` if it's possible to plant the flowers, `false` otherwise.
     */
    public boolean canPlaceFlowers(int[] flowerBed, int flowersToPlant) {
        boolean wasJustPlanted = false;

        for (int currentPlot : flowerBed) {
            if (currentPlot == 1 && wasJustPlanted) {
                flowersToPlant++;
            } else if (currentPlot == 1) {
                wasJustPlanted = true;
            } else if (!wasJustPlanted) {
                flowersToPlant--;
                wasJustPlanted = true;
            } else {
                wasJustPlanted = false;
            }
        }

        return flowersToPlant <= 0;
    }
}

File: src/main/kotlin/leetcode75/arraystring/CanPlaceFlowers.kt
package dev.artisra.leetcode75.arraystring

/**
 * A class providing a solution to the "Can Place Flowers" problem.
 *
 * Given a flowerbed represented as an array of 0s and 1s (0 means empty, 1 means not empty),
 * determines if it's possible to plant a given number of new flowers without planting them
 * in adjacent plots.
 * @see <a href="https://leetcode.com/problems/can-place-flowers/description/">Can Place Flowers on LeetCode</a>
 */
class CanPlaceFlowers {

    /**
     * Checks if it's possible to plant a given number of flowers in the flowerbed.
     *
     * @param flowerbed An array representing the flowerbed, where 0 means empty and 1 means not empty.
     * @param n The number of flowers to plant.
     * @return `true` if it's possible to plant all `n` flowers, `false` otherwise.
     */
    fun canPlaceFlowers(flowerbed: IntArray, n: Int): Boolean {
        var flowersToPlant = n
        var justPlanted = false
        for (currentPlot in flowerbed) {
            when {
                currentPlot == 1 && justPlanted -> flowersToPlant++
                currentPlot == 1 -> justPlanted = true
                !justPlanted -> {
                    flowersToPlant--
                    justPlanted = true
                }

                else -> justPlanted = false
            }
        }
        return flowersToPlant <= 0
    }
}
